### ISSUE TITLE
Filesystem support

### ISSUE THREAD

**Issue by btk** [expo/expo#108]
Hi, I decided to use expo for a little project, it was close to being flawless for 2 days of coding. I came at a point that I had to use file-system. But as I my understanding, there is no support for it, so I had to eject the app.

I wonder if I miss something or expo does not support it?

Also, is there any documentation for contributing or adding features?

Thank you, great tool btw.

**Comment by brentvatne** [comment]
Hi @btk! What do you need to do with the file system? We expose a few fs functions in a private API right now, @nikki93 has done some work on this and can give you more information if you provide details of your use case :)

**Comment by btk** [comment]
Hi @brentvatne , 
So basically, I want to download a zip file which possibly contains txt and image files, then unzip it (change its path), read the file, maybe edit it (in my case not important). I'm working on a book reader app, that you can download the book data from a server.

I know, when I introduce the zipping into the canvas, also expo needs to support it too. Which is not really a prior, in my state I can handle it without zipping.

Basically, seems like most of the features react-native-fs has would be useful.

**Comment by skevy** [comment]
Hi @btk! Filesystem support is something we have on our list, but don't yet have scheduled. We'll be sure to let you know when we release it!

**Comment by chourobin** [comment]
Just a question, are the files from ImagePicker cleaned up at some point? I am requesting filesystem support so that I can cleanup any files copied over from ImagePicker after uploading them.

**Comment by brentvatne** [comment]
@nikki93 ^

**Comment by thebigredgeek** [comment]
Any updates here?  This is blocking me :(

**Comment by nikki93** [comment]
@thebigredgeek Planning to make this a user-facing API for SDK 19 or later. Right now there is a 'private' API -- useable from JavaScript but not really documented yet. It's used by the Asset API here https://github.com/expo/expo-sdk/blob/master/src/Asset.js#L118-L144 for example. https://github.com/expo/expo/blob/master/ios/Exponent/Versioned/Modules/Api/EXFileSystem.m#L36-L120 shows what functions are available. Hit me up on slack.expo.io (I'm 'nikki' there) for more help on this if you have questions.

**Comment by dikaiosune** [comment]
Hi! This is in the works, but I'm going to move this to a [Canny feature request](https://expo.canny.io/feature-requests/p/filesystem-api) since that's where we're tracking enhancements now. Thanks!

**Comment by PavelPZ** [comment]
Hi! Thanks for this feature.
Will there be the possibility to write x read x seek binary blobs too?
I mentioned that https://github.com/hnq90/react-native-filesystem processes only UTF8 text files, without seek possibility.

**Comment by dikaiosune** [comment]
@PavelPZ please comment on the feature request linked in my above comment to make sure this feedback isn't missed!

**Comment by nikki93** [comment]
An initial version of this has been released. You can see the documentation for version 19 of the API here: https://docs.expo.io/versions/v19.0.0/sdk/filesystem.html

**Comment by EnriqueDev** [comment]
In SKD 20 the FileSystem API has a download method, but you can't append any headers to it. In my particular case, the download request must have an Authentication field in the http header in order to allow the user to download the file. Is there a way to do this?
